### DISCUSSION TITLE
How to restrict access to collections based on relationship keys

### DISCUSSION THREAD

**Discussion by NerminKarapandzic** [pocketbase/pocketbase#1580] [Q&A]
Let's imagine this situation:
We have 3 collections, clients, shops and products. Each client can own x shops. Each shop can have any number of products. 
I want to restrict adding products to shops you don't own, or adding products you don't own to your own shop.

So for example on product create api I would need to check that shop related to the product we're trying to insert belongs to the currently authenticated client. Can it be done?

**Comment by ganigeorgiev** [comment]
There are several ways to do it depending on the type of relation (single or multiple) you want to have between your data, but the simplest configuration would be something like:

| shops | `base` collection |
|:------|:------------------|
| name  | `text`            |

| products | `base` collection                    |
|:---------|:-------------------------------------|
| name     | `text`                               |
| shop     | `relation` (nonempty; maxSelect = 1) |

| users | `auth` collection                       |
|:------|:----------------------------------------|
| shops | `relation` (nonempty; maxSelect = null) |


Then for the `products` Create/Update API rules you can use:

```js
// Create rule, translates to:
// "Require an authenticated user"
// AND
// "At least one of the user's shops to match with the submitted product's shop"
@request.auth.id != "" && @request.auth.shops.id ?= @request.data.shop

// Update rule, translates to:
// "Require an authenticated user"
// AND
// "At least one of the user's shops to match with the already saved product's shop"
// AND
// "disallow shop change"
@request.auth.id != "" && @request.auth.shops.id ?= shop && @request.data.shop:isset == false
```

_You can also find some example access control patterns at the bottom of the https://pocketbase.io/docs/collections/#auth-collection page ._

**Reply by dschissler** [reply]
@ganigeorgiev Typically I would have a `shops` have a relation to the user, but it seems that there is another way to do it since the `users` can have a `shop` relation field that can have multiple values.

I just realized that I don't know how to interact with a multi-relation fields at all.  Could you point me to how to list, add and remove relation field entries with it in Go and JS?  Do you have to simply set all of the values each time, or does it do a difference?

**Reply by ganigeorgiev** [reply]
@dschissler

> Could you point me to how to list, add and remove relation field entries with it in Go and JS? 

This is not documented very well at the moment but in general it is the same as any other field value. Internally multi-`relation` values are stored as json array of relation ids (aka. the same as the returned representation).

To make it a little bit easier updating multi-`relation` fields, in v0.11.0 we have added support for `+` and `-` field body modifiers (you can find [some examples in the release notes](https://github.com/pocketbase/pocketbase/releases/tag/v0.11.0)).

Here are couple examples with the JS SDK:

```js
// set all relations at once
await pb.collection('example').create({ 'someRelField': ['id1', 'id2', 'id3'] });

// add a single (or multiple) value(s) to an existing rel array (v0.11.0+)
await pb.collection('example').update('RECORD_ID', { 'someRelField+': ['id4', 'id5'] });

// remove a single (or multiple) value(s) from an existing rel array from (v0.11.0+)
await pb.collection('example').update('RECORD_ID', { 'someRelField-': 'id2' });

// to "list" you can use expand similar to the single relation field
await pb.collection('example').getList(1, 40, { expand: 'someRelField' });
```

When working directly with the Go `models.Record`, you have to read/update the final/result slice field value, for example:

```go
// retrieve rel ids
ids := record.GetStringSlice("someRelField")

// add or remove rel ids
updatedIds := ...

// update the model field value
record.Set("someRelField", updatedIds)

// persist
app.Dao().SaveRecord(record)
```

**Reply by dschissler** [reply]
@ganigeorgiev I'll read this in detail later.  I have a question though if sqlite is able to index json values?

**Reply by ganigeorgiev** [reply]
SQLite supports index expressions so you can use it even with complex json object when combined with `json_extract`, but as any other column, it is not guaranteed that the query planner will choose it. My general recommendation if the performance starts to degrade would be to enable the `--debug` mode, copy paste the problematic query and run it in the sqlite3 cli tool with `EXPLAIN` and add indexes accordingly to the query planner (_adding indexes from the Admin UI will be handled as part of #345 and #544_).
